server/middleware/src/spaceSecurity.ts

```diff
@@ -0,0 +1,286 @@
+        }
+      } else if (!updateDoc.operations.private) {
+        this.removeSpace(updateDoc.objectId)
+      }
+    }
+
+    let space = this.privateSpaces[updateDoc.objectId]
+    if (space !== undefined) {
+      if (updateDoc.operations.members !== undefined) {
+        this.syncMembers(updateDoc.operations.members, space._id)
+      }
+      if (updateDoc.operations.$push?.members !== undefined) {
+        this.pushMembersHandle(updateDoc.operations.$push.members, space._id)
+      }
+
+      if (updateDoc.operations.$pull?.members !== undefined) {
+        this.pullMembersHandle(updateDoc.operations.$pull.members, space._id)
+      }
+      space = TxProcessor.updateDoc2Doc(space, updateDoc)
+    }
+  }
+
+  private handleRemove (tx: TxCUD<Space>): void {
+    const removeTx = tx as TxRemoveDoc<Space>
+    if (!this.storage.hierarchy.isDerived(removeTx.objectClass, core.class.Space)) return
+    if (removeTx._class !== core.class.TxCreateDoc) return
+    this.removeSpace(tx.objectId)
+  }
+
+  private async handleTx (ctx: SessionContext, tx: TxCUD<Space>): Promise<void> {
+    if (tx._class === core.class.TxCreateDoc) {
+      this.handleCreate(tx)
+    } else if (tx._class === core.class.TxUpdateDoc) {
+      await this.handleUpdate(ctx, tx)
+    } else if (tx._class === core.class.TxRemoveDoc) {
+      this.handleRemove(tx)
+    }
+  }
+
+  async getTargets (accounts: Ref<Account>[] | undefined): Promise<string[] | undefined> {
+    if (accounts === undefined) return
+    const users = await this.storage.modelDb.findAll(core.class.Account, { _id: { $in: accounts } })
+    return users.map((p) => p.email)
+  }
+
+  async tx (ctx: SessionContext, tx: Tx): Promise<TxMiddlewareResult> {
+    const h = this.storage.hierarchy
+    let targets: string[] | undefined
+
+    if (h.isDerived(tx._class, core.class.TxCUD)) {
+      const cudTx = tx as TxCUD<Doc>
+      const isSpace = h.isDerived(cudTx.objectClass, core.class.Space)
+      if (isSpace) {
+        await this.handleTx(ctx, cudTx as TxCUD<Space>)
+      }
+      const space = this.privateSpaces[tx.objectSpace]
+      if (space !== undefined) {
+        const account = await getUser(this.storage, ctx)
+        const allowed = this.allowedSpaces[account]
+        if (allowed === undefined || !allowed.includes(isSpace ? (cudTx.objectId as Ref<Space>) : tx.objectSpace)) {
+          throw new PlatformError(new Status(Severity.ERROR, platform.status.Forbidden, {}))
+        }
+        targets = await this.getTargets(this.privateSpaces[tx.objectSpace]?.members)
+      }
+    }
+
+    const res = await this.provideTx(ctx, tx)
+    return [res[0], res[1], mergeTargets(targets, res[2])]
+  }
+
+  override async findAll<T extends Doc>(
+    ctx: SessionContext,
+    _class: Ref<Class<T>>,
+    query: DocumentQuery<T>,
+    options?: FindOptions<T>
+  ): Promise<FindResult<T>> {
+    const findResult = this.storage.hierarchy.clone(
+      await this.provideFindAll(ctx, _class, query, options)
```

Comment:
Why not {...query, space: {$in: allowed}}?

Reply:
Allowed contains only private spaces

Reply:
I've mean it should be faster to filter on Mongo, not on server. allowed is concept of all spaces allowed.